### PR TITLE
Rake task and job

### PR DESCRIPTION
**Pull request by brendandeere**
Actionable subscriptions and installments can now be processed with the
rake task:

`bundle exec rake solidus_subscriptions:process`

### CODE REVIEW

File: lib/solidus_subscriptions/processor.rb [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+# the consolidated installment class.
+#
+# This class generates `ProcessInstallmentsJob`s
+module SolidusSubscriptions
+  class Processor
+    class << self
+      # Find all actioanble subscriptions and intallments, group them together
+      # by user, and schedule a processing job for the group as a batch
+      def run
+        batched_users_to_be_procesed.each { |batch| new(batch).build_jobs }
```
Comment (alexblackie) [inline lib/solidus_subscriptions/processor.rb]: processed*

File: lib/solidus_subscriptions/processor.rb [hunk 2]
```diff
@@ -0,0 +1,91 @@
+# This class is responsible for finding subscritpions and installments
```
Comment (alexblackie) [inline lib/solidus_subscriptions/processor.rb]: subscriptions*

File: lib/solidus_subscriptions/processor.rb [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+          ).
+          distinct.
+          find_in_batches
+      end
+    end
+
+    # @return [Array<Spree.user_class>]
+    attr_reader :users
+
+    # Get a new instance of the SolidusSubscritpions::Processor
```
Comment (alexblackie) [inline lib/solidus_subscriptions/processor.rb]: SolidusSubscriptions*

File: lib/solidus_subscriptions/processor.rb [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    #
+    # @param :users [Array<Spree.user_class>] A list of users with actionable
```
Comment (alexblackie) [inline lib/solidus_subscriptions/processor.rb]: `:users` should not be a symbol, just a keyword in YARD. `@param users [...`

File: lib/solidus_subscriptions/processor.rb [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    def retry_installments
+      @failed_installments ||= Installment.
+        actionable.
+        includes(:subscription).
+        where(solidus_subscriptions_subscriptions: { user_id: user_ids }).
+        group_by { |i| i.subscription.user_id }
+    end
+
+    def installments(user)
+      return @installments[user.id] if @installments[user.id]
```
Comment (alexblackie) [inline lib/solidus_subscriptions/processor.rb]: could this not just be an `||=` on the line below?

File: lib/solidus_subscriptions/testing_support/factories/subscription_factory.rb [hunk 6]
```diff
@@ -9,5 +12,18 @@
 
       line_item { build :subscription_line_item, *line_item_traits }
     end
+
+    trait :actionable do
+      with_line_item
+      actionable_date { Time.zone.now.yesterday }
+    end
+
+    trait :inactionable do
```
Comment (alexblackie) [inline lib/solidus_subscriptions/testing_support/factories/subscription_factory.rb]: so minor thing, but this has some weird ambiguity, I think. I read this the first time as "able to be inactioned", as opposed to "not actionable"... Changing to `not_actionable` might be clearer

File: lib/solidus_subscriptions/processor.rb (hunk 1 above)
Comment (brendandeere) [inline lib/solidus_subscriptions/processor.rb]: nice :P

File: lib/solidus_subscriptions/processor.rb (hunk 5 above)
Comment (brendandeere) [inline lib/solidus_subscriptions/processor.rb]: Yeah, Originally there were more lines here. I pulles some stuff out into smaller methods

File: lib/solidus_subscriptions/testing_support/factories/subscription_factory.rb (hunk 6 above)
Comment (brendandeere) [inline lib/solidus_subscriptions/testing_support/factories/subscription_factory.rb]: :+1:

**Comment by brendandeere** [comment]
@alexblackie Fixed the things

File: lib/solidus_subscriptions/processor.rb [hunk 7]
```diff
@@ -0,0 +1,90 @@
+# This class is responsible for finding subscriptions and installments
+# which need to be processed. It will groups them together by user and attempts
```
Comment (alexblackie) [inline lib/solidus_subscriptions/processor.rb]: >It will group

File: lib/solidus_subscriptions/processor.rb [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+# to process them together.
+#
+# This class passes the reponsibility of actually creating the order off onto
+# the consolidated installment class.
+#
+# This class generates `ProcessInstallmentsJob`s
+module SolidusSubscriptions
+  class Processor
+    class << self
+      # Find all actioanble subscriptions and intallments, group them together
```
Comment (alexblackie) [inline lib/solidus_subscriptions/processor.rb]: actionable

File: lib/solidus_subscriptions/processor.rb [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+          where(
+            Subscription.actionable.arel.constraints.reduce(:and).
+              or(Installment.actionable.arel.constraints.reduce(:and))
+          ).
+          distinct.
+          find_in_batches
+      end
+    end
+
+    # @return [Array<Spree.user_class>]
```
Comment (alexblackie) [inline lib/solidus_subscriptions/processor.rb]: I believe YARD has a syntax for documenting accessors at the class level

**Comment by brendandeere** [comment]
@alexblackie Fixed your comments :)

**Pull request merged**